src/LinAlg/hiopLinSolverSparseGinkgo.cpp

```diff
@@ -0,0 +1,376 @@
+          std::sort(jCol_+kRowPtr_[i],jCol_+kRowPtr_[i+1]);
+        }
+      }
+
+      delete[] nnz_each_row_tmp;
+    }
+
+    auto val_array = gko::Array<double>::view(exec, nnz_, kVal_);
+    auto row_ptrs = gko::Array<int>::view(exec, n_ + 1, kRowPtr_);
+    auto col_idxs = gko::Array<int>::view(exec, nnz_, jCol_);
+    auto mtx = gko::share(gko::matrix::Csr<double, int>::create(exec, gko::dim<2>{n_, n_}, val_array, col_idxs, row_ptrs));
+    return mtx;
+}
+
+
+void update_matrix(hiopMatrixSparse* M_, std::shared_ptr<gko::matrix::Csr<double, int>> mtx, int *index_covert_CSR2Triplet_, int * index_covert_extra_Diag2CSR_) {
+    int n_ = mtx->get_size()[0];
+    int nnz_= mtx->get_num_stored_elements();
+    auto values = mtx->get_values();
+    int rowID_tmp{0};
+    for(int k=0;k<nnz_;k++){
+    values[k] = M_->M()[index_covert_CSR2Triplet_[k]];
+    }
+    for(int i=0;i<n_;i++){
+    if(index_covert_extra_Diag2CSR_[i] != -1)
+        values[index_covert_extra_Diag2CSR_[i]] += M_->M()[M_->numberOfNonzeros()-n_+i];
+    }
+}
+
+
+std::shared_ptr<gko::LinOpFactory> setup_solver_factory(std::shared_ptr<const gko::Executor> exec,
+        std::shared_ptr<gko::matrix::Csr<double, int>> mtx) {
+
+    auto preprocessing_fact = gko::share(gko::reorder::Mc64<double, int>::build().on(exec));
+    auto preprocessing = gko::share(preprocessing_fact->generate(mtx));
+
+    auto lu_fact = gko::share(gko::factorization::Glu<double, int>::build_reusable()
+            .on(exec, mtx.get(), preprocessing.get()));
+    auto inner_solver_fact = gko::share(gko::preconditioner::Ilu<>::build()
+            .with_factorization_factory(lu_fact)
+            .on(exec));
+    auto solver_fact = gko::share(gko::solver::Gmres<>::build()
+            .with_criteria(
+                gko::stop::Iteration::build()
+                    .with_max_iters(200u)
+                    .on(exec),
+                gko::stop::ResidualNorm<>::build()
+                    .with_baseline(gko::stop::mode::absolute)
+                    .with_reduction_factor(1e-8)
+                    .on(exec))
+            .with_krylov_dim(10u)
+            .with_preconditioner(inner_solver_fact)
+            .on(exec));
+
+    auto reusable_factory = gko::share(gko::solver::ScaledReordered<>::build()
+            .with_solver(solver_fact)
+            .with_reordering(preprocessing)
+            .on(exec));
+    return reusable_factory;
+}
+
+
+}
+
+
+  hiopLinSolverIndefSparseGinkgo::hiopLinSolverIndefSparseGinkgo(const int& n, const int& nnz, hiopNlpFormulation* nlp)
+    : hiopLinSolverSymSparse(n, nnz, nlp),
+      index_covert_CSR2Triplet_{nullptr},index_covert_extra_Diag2CSR_{nullptr},
+      n_{n}, nnz_{0}
+  {}
+
+  hiopLinSolverIndefSparseGinkgo::~hiopLinSolverIndefSparseGinkgo()
+  {
+    if(index_covert_CSR2Triplet_)
+      delete [] index_covert_CSR2Triplet_;
+    if(index_covert_extra_Diag2CSR_)
+      delete [] index_covert_extra_Diag2CSR_;
```

Comment:
[Deleting a nullptr is a safe no-op, here's an example.](https://godbolt.org/z/Ks8oh55W9)

[From the standard:](https://en.cppreference.com/w/cpp/language/delete)

> If expression evaluates to a null pointer value, no destructors are called, and the deallocation function may or may not be called (it's unspecified), but the default deallocation functions are guaranteed to do nothing when passed a null pointer.

```suggestion
    delete [] index_covert_CSR2Triplet_;
    delete [] index_covert_extra_Diag2CSR_;
```

We do however have the pattern `if (ptr) delete ptr;` all over hiop, so it's up to @cnpetra if we should keep up the pattern.

Reply:
I am for using `delete ptr;` only as you pointe out. btw, I started using it quite some time ago.